File: Niveau 4/Arbres/Fibre optique/main.cpp
#include <iostream>
#include <vector>
using Vector = std::vector<std::vector<int>>;
template <typename T>
T max_f(T count, T max)
{
  return (count > max) ? count : max;
}
template <typename T>
T min_f(T count, T max)
{
  return (count < max) ? count : max;
}
void dfs(int node, int parent, Vector &adj, std::vector<int> &subtree_size, int &max_deco)
{
  subtree_size [node] = 1; //initialisation
  for (int voisin : adj [node])
  {
    if (voisin == parent)
      continue;
    dfs(voisin, node, adj, subtree_size, max_deco);
    subtree_size [node] += subtree_size [voisin]; //ajoute la taille du sous arbre
  }
}
int main()
{
  int n;
  std::cin >> n;
  Vector adj(n);
  for (int i {0}; i < n-1; i++)
  {
    int a, b;
    std::cin >> a >> b;
    adj [a].push_back(b);
    adj [b].push_back(a);
  }
/*   std::cout << "--------------\n";
  for (int i {0}; i < n; i++)
  {
    std::cout << "PC " << i;
    for (int j {0}; j < (int) adj [i].size(); j++)
    {
      std::cout << " : " << adj [i][j];
    }
    std::cout << '\n';
  } */
  std::vector<int> sub_tree(n, 0);
  int max_deco {0};
  dfs(0, -1, adj, sub_tree, max_deco);
  for (int i {0}; i < n; i++)
  {
    for (int voisin : adj [i])
    {
      if (sub_tree [voisin] < sub_tree [i])
      {
        int taille1 {sub_tree [voisin]};
        int taille2 {n - taille1};
        max_deco = max_f(max_deco, min_f(taille1, taille2));
      }
    }
  }
  std::cout << max_deco << '\n';
  return 0;
}

File: Niveau 4/Arbres/Fibre optique/main.py
import sys
def max_f(count, max_val):
    return count if count > max_val else max_val
def min_f(count, max_val):
    return count if count < max_val else max_val
def dfs(node, parent, adj, subtree_size):
    subtree_size[node] = 1  # initialisation
    for voisin in adj[node]:
        if voisin == parent:
            continue
        dfs(voisin, node, adj, subtree_size)
        subtree_size[node] += subtree_size[voisin]  # ajoute la taille du sous-arbre
def main():
    input = sys.stdin.readline
    n = int(input().strip())
    adj = [[] for _ in range(n)]
    
    for _ in range(n - 1):
        a, b = map(int, input().split())
        adj[a].append(b)
        adj[b].append(a)
    
    subtree_size = [0] * n
    max_deco = 0
    
    dfs(0, -1, adj, subtree_size)
    
    for i in range(n):
        for voisin in adj[i]:
            if subtree_size[voisin] < subtree_size[i]:
                taille1 = subtree_size[voisin]
                taille2 = n - taille1
                max_deco = max_f(max_deco, min_f(taille1, taille2))
    
    print(max_deco)
if __name__ == "__main__":
    sys.setrecursionlimit(200 * 1000)
    main()
